### PR TITLE
Allow microvm configuration from a single json passed as command line parameter

### PR DESCRIPTION
**Pull request by lauralt**
Issue #923 , if available:

Description of changes:
Added extra parameter 'config-file' to Firecracker, which allows starting a microvm without sending API requests. User has to send  in the start command a file which stores a single json, that contains the configuration for all of the resources used by the microvm. Also, added an integration test which successfully starts a microvm with a configuration file sent as command line parameter and unit tests.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: vmm/src/lib.rs [hunk 1]
```diff
@@ -3388,6 +3460,41 @@ mod tests {
...
+    fn test_configure_vmm_from_json() {
+        let mut vmm = create_vmm_object(InstanceState::Uninitialized);
+
+        let kernel_file = NamedTempFile::new().unwrap();
+        let rootfs_file = NamedTempFile::new().unwrap();
+
+        let json = String::from(
+            r#"{
+                    "boot_source":{
+                        "kernel_image_path": ""#,
```
Comment (aghecenco) [inline vmm/src/lib.rs]: Why not `format!` the string? The alignment is weird and it's not very readable.

File: vmm/src/lib.rs [hunk 2]
```diff
@@ -712,6 +713,18 @@ struct KernelConfig {
     cmdline_addr: GuestAddress,
 }
 
+/// Used for configuring a vmm from one single json passed to the Firecracker process
```
[the remark this answers is not in the document]
Comment (lauralt) [reply vmm/src/lib.rs]: Done.

File: vmm/src/lib.rs [hunk 3]
```diff
@@ -712,6 +713,18 @@ struct KernelConfig {
...
+#[derive(Deserialize)]
+pub struct VmmConfig {
+    boot_source: BootSourceConfig,
+    block_devices: Vec<BlockDeviceConfig>,
+    net_devices: Option<Vec<NetworkInterfaceConfig>>,
```
[the remark this answers is not in the document]
Comment (lauralt) [reply vmm/src/lib.rs]: If I'd use an empty Vec, if user doesn't want to insert any network interface, he must however add in the configuration json, an empty json for those devices, something like:
network-interfaces": [
        ], 
which IMO doesn't look nice.

File: vmm/src/lib.rs [hunk 4]
```diff
@@ -2040,6 +2053,48 @@ impl Vmm {
...
+
+    fn configure_from_json(
+        &mut self,
+        config_json: String,
+    ) -> std::result::Result<(), VmmActionError> {
+        let vmm_config = match serde_json::from_slice::<VmmConfig>(config_json.as_bytes()) {
+            Ok(config) => config,
+            Err(e) => {
+                error!("Invalid json: {}", e);
+                process::exit(i32::from(FC_EXIT_CODE_GENERIC_ERROR));
```
[the remark this answers is not in the document]
Comment (lauralt) [reply vmm/src/lib.rs]: Done

File: vmm/src/lib.rs [hunk 5]
```diff
@@ -2105,13 +2160,30 @@ pub fn start_vmm_thread(
...
+                match vmm.configure_from_json(json) {
+                    Ok(()) => {
+                        info!("Successfully configured VMM from one single json");
+                    }
+                    Err(e) => {
+                        error!(
+                            "Setting configuration for VMM from one single json failed: {}",
+                            e
+                        );
+                        process::exit(i32::from(FC_EXIT_CODE_GENERIC_ERROR));
```
[the remark this answers is not in the document]
Comment (lauralt) [reply vmm/src/lib.rs]: Done

File: jailer/src/env.rs [hunk 6]
```diff
@@ -368,6 +381,25 @@ mod tests {
         let gid = "1002";
         let chroot_base = "/";
         let netns = "zzzns";
+        let config_json = r#"{
+           "boot_source":{
```
[the remark this answers is not in the document]
Comment (lauralt) [reply jailer/src/env.rs]: I agree, I renamed the resources with the names from .yaml file. Please, take a look!

File: tests/framework/jailer.py [hunk 7]
```diff
@@ -62,7 +86,7 @@ def __del__(self):
         """Cleanup this jailer context."""
         self.cleanup()
 
-    def construct_param_list(self):
+    def construct_param_list(self, config_json):
```
[the remark this answers is not in the document]
Comment (lauralt) [reply tests/framework/jailer.py]: Yes, it looks better now.

File: tests/framework/jailer.py [hunk 8]
```diff
@@ -113,16 +139,17 @@ def chroot_path(self):
...
-    def jailed_path(self, file_path, create=False):
+    def jailed_path(self, file_path, create=False, json_start=False):
         """Create a hard link owned by uid:gid.
 
         Create a hard link to the specified file, changes the owner to
         uid:gid, and returns a path to the link which is valid within the jail.
         """
         file_name = os.path.basename(file_path)
         global_p = os.path.join(self.chroot_path(), file_name)
+        if json_start:
```
[the remark this answers is not in the document]
Comment (lauralt) [reply tests/framework/jailer.py]: To create the resources needed for configuring the microvm, chroot_path should exist before calling spawn(). This path used to be created by the jailer, but we have to enforce that creation before that call of spawn().

File: tests/integration_tests/functional/test_cmd_line_start.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (lauralt) [reply tests/integration_tests/functional/test_cmd_line_start.py]: Done

File: vmm/src/lib.rs (hunk 1 above)
Comment (lauralt) [reply vmm/src/lib.rs]: I agree, it looks better.

File: vmm/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tamionv) [reply vmm/src/lib.rs]: Hey, I think you can fix this with these attributes:
https://github.com/dtolnay/serde-yaml/issues/32

File: vmm/src/lib.rs [hunk 10]
```diff
@@ -712,6 +713,18 @@ struct KernelConfig {
...
+/// Used for configuring a vmm from one single json passed to the Firecracker process
+#[derive(Deserialize)]
+pub struct VmmConfig {
+    boot_source: BootSourceConfig,
+    block_devices: Vec<BlockDeviceConfig>,
+    net_devices: Option<Vec<NetworkInterfaceConfig>>,
+    logger: Option<LoggerConfig>,
+    machine_config: Option<VmConfig>,
+    #[cfg(feature = "vsock")]
+    vsock_devices: Option<Vec<VsockDeviceConfig>>,
```
[the remark this answers is not in the document]
Comment (tamionv) [reply vmm/src/lib.rs]: This can also be fixed with these attributes:
https://github.com/dtolnay/serde-yaml/issues/32

File: vmm/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tamionv) [reply vmm/src/lib.rs]: (as in:
```
#[serde(rename = "network-interfaces", skip_serializing_if = "Vec::is_empty", default)]
net_devices: Vec<NetworkInterfaceConfig>,
```

File: vmm/src/lib.rs [hunk 11]
```diff
@@ -2064,8 +2068,7 @@ impl Vmm {
         &mut self,
         config_json: String,
     ) -> std::result::Result<(), VmmActionError> {
-        let body: Chunk = Chunk::from(config_json);
-        let vmm_config = match serde_json::from_slice::<VmmConfig>(&body) {
+        let vmm_config = match serde_json::from_slice::<VmmConfig>(config_json.as_bytes()) {
```
Comment (tamionv) [inline vmm/src/lib.rs]: Could you do something like:
```
serde_json::from_slice::<VmmConfig>(config_json.as_bytes()).unwrap_or_else(|e| {
    error!("Invalid json: {}", e);
    process::exit(i32::from(FC_EXIT_CODE_INVALID_JSON));
}
```
? It feels slightly more expressive (i.e. rather than having to parse the match, you immediately see that we take the value of the config_json, propagating any errors generated by from_slice in some way.

File: vmm/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lauralt) [reply vmm/src/lib.rs]: Worked :thumbsup:. Also, removed `skip_serializing_if = "Vec::is_empty"`, as we're using `VmmConfig` only for deserialization.

File: vmm/src/lib.rs (hunk 11 above)
Comment (lauralt) [reply vmm/src/lib.rs]: Done

File: vmm/src/lib.rs [hunk 12]
```diff
@@ -2040,6 +2063,48 @@ impl Vmm {
...
+    ) -> std::result::Result<(), VmmActionError> {
+        let vmm_config = match serde_json::from_slice::<VmmConfig>(config_json.as_bytes()) {
+            Ok(config) => config,
+            Err(e) => {
+                error!("Invalid json: {}", e);
+                process::exit(i32::from(FC_EXIT_CODE_INVALID_JSON));
+            }
+        };
+
+        if let Some(logger) = vmm_config.logger {
```
[the remark this answers is not in the document]
Comment (lauralt) [reply vmm/src/lib.rs]: I prefer to leave it as it is, I find it more suggestive that way.

File: tests/integration_tests/security/demo_seccomp/Cargo.lock [hunk 13]
```diff
@@ -1,3 +1,5 @@
+# This file is automatically @generated by Cargo.
```
Comment (aghecenco) [inline tests/integration_tests/security/demo_seccomp/Cargo.lock]: How'd this end up in the PR? I think these Cargo.lock comments appeared after a rustc upgrade, but this crate shouldn't be built as part of the "normal" build process; is it a leftover from running tests outside the container? In either case, please revert & remove it from this PR

File: tests/framework/microvm.py [hunk 14]
```diff
@@ -207,9 +211,10 @@ def memory_events_queue(self, queue):
         """Set the memory usage events queue."""
         self._memory_events_queue = queue
 
-    def create_jailed_resource(self, path):
+    def create_jailed_resource(self, path, json_start=False):
```
Comment (aghecenco) [inline tests/framework/microvm.py]: I find this counterintuitive, a helper function that hardlinks stuff shouldn't care about how Firecracker is started. I'd rename the parameter to `create_jail` (or something similar) instead.

Comment (lauralt) [reply tests/framework/microvm.py]: Renamed it.

File: jailer/src/env.rs [hunk 15]
```diff
@@ -295,20 +299,23 @@ impl Env {
...
-                .arg(format!("--api-sock=/{}", socket_file_name))
-                .stdin(Stdio::inherit())
-                .stdout(Stdio::inherit())
-                .stderr(Stdio::inherit())
-                .uid(self.uid())
-                .gid(self.gid())
-                .exec(),
-        ))
+        let mut command = Command::new(chroot_exec_file);
+        let init_command = command
```
Comment (acatangiu) [inline jailer/src/env.rs]: nit: do you need to use another variable (`init_command`)?

File: vmm/src/lib.rs [hunk 16]
```diff
@@ -3388,6 +3462,40 @@ mod tests {
             .is_some());
     }
 
+    #[test]
+    fn test_configure_vmm_from_json() {
```
Comment (acatangiu) [inline vmm/src/lib.rs]: When testing `configure_from_json`, please try to test all branches of the function. Use valid and invalid resources for each resource and `assert` on the expected errors on the invalid types.

Unfortunately we can't actually boot a microVM (see commit `1754f991`) so we might not have any way to test this function unless you decouple the configuration code from the `start_microvm()` so that we can at least test the configuration part.

**Comment by alexandruag** [comment]
> I'm a bit worried that we're missing a good opportunity to share the relevant parts with our API path so that we also support single API call config.
> I believe clients that still need the API (e.g. for post-boot operations) would be highly grateful if they could do the single API call using their existing path (swagger-generated API client) instead of having to manually create JSONs and pass them as command-line args.

That's a good point, but a bit outside the scope of this PR. The opportunity to build on top of this remains, and we can prioritize it if there's a real customer need, or pick it up as follow-up work at some point.

**Comment by acatangiu** [comment]
I agree that we needn't target the single API call config in this PR, but I think that @lauralt has a good opportunity to do that as well (in subsequent PR/PRs) - it has real customer value.

File: jailer/src/env.rs (hunk 15 above)
Comment (lauralt) [reply jailer/src/env.rs]: Yes, in order to increase the lifetime of the Command object created by `new() `function.

File: vmm/src/lib.rs (hunk 16 above)
Comment (lauralt) [reply vmm/src/lib.rs]: Ok, i decoupled the configuration code and added more unit tests. Now `configure_from_json()` is fully covered, but the code coverage decreased due to the fact that the newly added lines in `start_vmm_thread()` function are hard to test.

File: jailer/src/env.rs (hunk 15 above)
Comment (tamionv) [reply jailer/src/env.rs]: Wouldn't something like this work also?
```
        let mut command = Command::new(chroot_exec_file);

        command
            .arg(format!("--id={}", self.id))
            // ...
            .gid(self.gid());

        if let Some(json) = self.config_json {
            command.arg(format!("--vmm-config={}", json));
        }
        Err(Error::Exec(command.exec()))
```

Comment (lauralt) [reply jailer/src/env.rs]: Yep, this works too, didn't try that.

File: CHANGELOG.md [hunk 17]
```diff
@@ -7,6 +7,11 @@
 - New device: virtio-vsock, backed by Unix domain sockets (GitHub issue #650).
   See `docs/vsock.md`.
 
+### Added
+
+- New parameter, `vmm-config`, which represents a JSON that can be used for configuring
```
Comment (acatangiu) [inline CHANGELOG.md]: nit: Be a bit more specific: "New command-line parameter for both `firecracker` and `jailer`, named `vmm-config`, bla."

File: CHANGELOG.md [hunk 18]
```diff
@@ -7,6 +7,11 @@
 - New device: virtio-vsock, backed by Unix domain sockets (GitHub issue #650).
   See `docs/vsock.md`.
 
+### Added
+
+- New parameter, `vmm-config`, which represents a JSON that can be used for configuring 
+  a microVM without sending any API requests.
```
Comment (acatangiu) [inline CHANGELOG.md]: Clearly specify that it configures **and starts** the microVM.

File: docs/getting-started.md [hunk 19]
```diff
@@ -242,6 +242,33 @@ curl --unix-socket /tmp/firecracker.socket -i  \
...
+by passing the parameter `vmm-config` to the Firecracker process. The command for starting 
+Firecracker with this option will look like this:
+
+```bash
+./firecracker --api-sock /tmp/firecracker.socket --vmm-config "JSON_BODY"
+```    
+
+"JSON_BODY" should contain the entire configuration for all of the microVM's resources. 
+You **must** provide in your JSON the configuration for the guest kernel and rootfs, as 
+these are mandatory, but all of the other resources are optional, so it's your choice if
```
Comment (acatangiu) [inline docs/getting-started.md]: Clearly specify that using this configuration method will also start the microVM with said configuration, so that all desired pre-boot configurable resources need to be specified.

File: docs/jailer.md [hunk 20]
```diff
@@ -38,6 +39,8 @@ jailer --id <id> \
     Firecracker.
   - 2 (default): advanced filtering. This adds further checks on some of the
     parameters of the allowed syscalls.
+- `vmm_config` represents the json that can be used for configuring the
```
Comment (acatangiu) [inline docs/jailer.md]:
```suggestion
- `vmm_config` represents the json that can be used for configuring and starting the 
```

File: CHANGELOG.md [hunk 21]
```diff
@@ -1,5 +1,13 @@
 # Changelog
 
+### Added
```
Comment (acatangiu) [inline CHANGELOG.md]: Please add a subsection called `[Unreleased]` above.

It should look like:

```
# Changelog

## [Unreleased]

### Added

Blablah blah
```

File: Cargo.lock [hunk 22]
```diff
@@ -969,6 +969,7 @@ dependencies = [
  "epoll 4.0.1 (registry+https://github.com/rust-lang/crates.io-index)",
  "fc_util 0.1.0",
  "futures 0.1.18 (registry+https://github.com/rust-lang/crates.io-index)",
+ "hyper 0.11.16 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (acatangiu) [inline Cargo.lock]: Is this supposed to be here? Why is `vmm` dependent on `hyper`?

File: vmm/src/lib.rs [hunk 23]
```diff
@@ -748,6 +755,18 @@ struct KernelConfig {
...
 
+/// Used for configuring a vmm from one single json passed to the Firecracker process.
+#[derive(Deserialize)]
+pub struct VmmConfig {
+    boot_source: BootSourceConfig,
+    block_devices: Vec<BlockDeviceConfig>,
+    net_devices: Option<Vec<NetworkInterfaceConfig>>,
+    logger: Option<LoggerConfig>,
+    machine_config: Option<VmConfig>,
+    #[cfg(feature = "vsock")]
```
Comment (acatangiu) [inline vmm/src/lib.rs]: This is no longer the case, there is no more `vsock` feature.

File: vmm/src/lib.rs [hunk 24]
```diff
@@ -2031,6 +2050,49 @@ impl Vmm {
...
+        }
+        if let Some(net_devices) = vmm_config.net_devices {
+            for net_config in net_devices.into_iter() {
+                self.insert_net_device(net_config)?;
+            }
+        }
+        if let Some(machine_config) = vmm_config.machine_config {
+            self.set_vm_configuration(machine_config)?;
+        }
+        #[cfg(feature = "vsock")]
```
Comment (acatangiu) [inline vmm/src/lib.rs]: Same here, no `#[cfg(feature = "vsock")]`

File: CHANGELOG.md [hunk 25]
```diff
@@ -1,6 +1,14 @@
...
-## Unreleased
+## [Unreleased]
+
+### Added
+
+- New command-line parameter for `firecracker`, named `config-file`, which 
+  represents the path to a file that contains a JSON which can be used for 
+  configuring and starting a microVM without sending any API requests.
+- The jailer adheres to the "end of command options" convention, meaning 
+  all parameters specified after -- are forwarded verbatim to Firecracker.
```
Comment (acatangiu) [inline CHANGELOG.md]:
```suggestion
  all parameters specified after `--` are forwarded verbatim to Firecracker.
```

File: FAQ.md [hunk 26]
```diff
@@ -196,3 +196,10 @@ Possible mitigations are:
   kernel so as to use `vmalloc` instead of `kmalloc` for them.
 - Reduce memory pressure on the host.
 
+### How can I configure and start a microVM without sending API calls?
+
+Passing an optional command line parameter, `config-file`, to the Firecracker 
+process allows this type of configuration. This parameter must be the path to a 
+file that contains the JSON specification that will be used to configure and start 
+the microVM. One example of such file, named `vm_config.json`, can be found in 
+`tests/framework`.
```
Comment (acatangiu) [inline FAQ.md]: nit: maybe add the full relative path here: `tests/framework/vm_config.json`.

File: FAQ.md [hunk 27]
```diff
@@ -196,3 +196,10 @@ Possible mitigations are:
   kernel so as to use `vmalloc` instead of `kmalloc` for them.
 - Reduce memory pressure on the host.
 
+### How can I configure and start a microVM without sending API calls?
+
+Passing an optional command line parameter, `config-file`, to the Firecracker
```
Comment (acatangiu) [inline FAQ.md]:
```suggestion
Passing an optional command line parameter, `--config-file`, to the Firecracker 
```

File: docs/getting-started.md [hunk 28]
```diff
@@ -242,6 +242,29 @@ curl --unix-socket /tmp/firecracker.socket -i  \
     }'
 ```
 
+#### Configuring the microVM without sending API requests
+
+If you'd like to boot up a guest machine without using the API socket, you can do that 
+by passing the parameter `config-file` to the Firecracker process. The command for
```
Comment (acatangiu) [inline docs/getting-started.md]:
```suggestion
by passing the parameter `--config-file` to the Firecracker process. The command for 
```

File: docs/getting-started.md [hunk 29]
```diff
@@ -242,6 +242,29 @@ curl --unix-socket /tmp/firecracker.socket -i  \
...
+starting Firecracker with this option will look like this:
+
+```bash
+./firecracker --api-sock /tmp/firecracker.socket --config-file 
+<path_to_the_configuration_file>
+```    
+
+`path_to_the_configuration_file` should represent the path to a file that contains a 
+JSON which stores the entire configuration for all of the microVM's resources. You 
+**must** provide in your JSON the configuration for the guest kernel and rootfs, as
```
Comment (acatangiu) [inline docs/getting-started.md]: nit: I'd rephrase to:
> The JSON **must** contain the configuration ... blah

File: docs/getting-started.md [hunk 30]
```diff
@@ -242,6 +242,29 @@ curl --unix-socket /tmp/firecracker.socket -i  \
...
+
+`path_to_the_configuration_file` should represent the path to a file that contains a 
+JSON which stores the entire configuration for all of the microVM's resources. You 
+**must** provide in your JSON the configuration for the guest kernel and rootfs, as 
+these are mandatory, but all of the other resources are optional, so it's your choice 
+if you want to configure them or not. Because using this configuration method will 
+also start the microVM, you need to specify all desired pre-boot configurable resources 
+in that JSON. The names of the resources are the ones from the `firecracker.yaml` file 
+and the names of their fields are the same that are used in API requests. 
+You can find an example of configuration file, named `vm_config.json`, in `tests/framework`.
```
Comment (acatangiu) [inline docs/getting-started.md]:
```suggestion
You can find an example of configuration file, at `tests/framework/vm_config.json`.
```

File: jailer/src/env.rs [hunk 31]
```diff
@@ -52,7 +52,7 @@ pub struct Env {
     seccomp_level: u32,
     start_time_us: u64,
     start_time_cpu_us: u64,
-    config_json: Option<String>,
+    extra_args: Option<Vec<String>>,
```
[the remark this answers is not in the document]
Comment (lauralt) [reply jailer/src/env.rs]: Since `values_of` returns an Option, I'd prefer to leave it as it is, if you don't strongly disagree with this.

[the remark this answers is not in the document]
Comment (acatangiu) [reply jailer/src/env.rs]: You can just add `unwrap_or()` like so:
```
        let extra_args = args
            .values_of("extra-args")
            .map(|v| v.collect::<Vec<&str>>())
            .map(|v| v.iter().map(|s| s.to_string()).collect()).
            .unwrap_or(vec![]);
```

But I won't block the PR for this, it's ultimately your choice.

[the remark this answers is not in the document]
Comment (lauralt) [reply jailer/src/env.rs]: Done, updated with empty `Vec`.

File: jailer/src/env.rs [hunk 32]
```diff
@@ -52,6 +52,7 @@ pub struct Env {
     seccomp_level: u32,
     start_time_us: u64,
     start_time_cpu_us: u64,
+    extra_args: Vec<String>,
```
Comment (tamionv) [inline jailer/src/env.rs]: Nit: I would name this something like `forwarded_args`, but it's up to you.

File: jailer/src/env.rs [hunk 33]
```diff
@@ -112,6 +113,12 @@ impl Env {
             .parse::<u32>()
             .map_err(Error::SeccompLevel)?;
 
+        let extra_args = args
```
Comment (tamionv) [inline jailer/src/env.rs]: I suggest changing this to `args.values_of("...").iter().flatten().map(to_string).collect()` or something like that.

File: src/main.rs [hunk 34]
```diff
@@ -142,6 +150,12 @@ fn main() {
             .expect("'start-time-cpu_us' parameter expected to be of 'u64' type.")
     });
 
+    let vmm_config_json = if let Some(path) = cmd_arguments.value_of("config-file") {
```
Comment (tamionv) [inline src/main.rs]: I would change this to something like this:
```
let vmm_config_json = cmd_arguments.value_of("config-file")
    .map(fs::read_to_string)
    .map(|x| x.expect("Unable to open or read from the configuration file")
```

File: jailer/src/env.rs (hunk 33 above)
Comment (acatangiu) [reply jailer/src/env.rs]: Or
```
        let extra_args = args
            .values_of("extra-args")
            .iter()
            .flat_map(to_string)
            .collect();
```

Comment (lauralt) [reply jailer/src/env.rs]: Thanks for the suggestions, I’ve started from them and managed to find a working solution which is now in the code.

File: jailer/src/env.rs (hunk 32 above)
Comment (lauralt) [reply jailer/src/env.rs]: I like `extra_args` more.

File: src/main.rs (hunk 34 above)
Comment (lauralt) [reply src/main.rs]: Done
